algorithms/src/traits/snark.rs

```diff
@@ -20,29 +20,48 @@ use snarkvm_utilities::{FromBytes, ToBytes};
 use rand::Rng;
 use std::fmt::Debug;
 
+/// Defines a trait that describes preparing from an unprepared version to a prepare version
+pub trait Prepare<T> {
+    fn prepare(&self) -> T;
+}
+
 pub trait SNARK {
     type AllocatedCircuit;
     type Circuit;
-    type PreparedVerifyingKey: Clone + From<Self::ProvingKey> + From<Self::VerifyingKey>;
+    type PreparedVerifyingKey: Clone;
     type Proof: Clone + Debug + ToBytes + FromBytes;
     type ProvingKey: Clone + ToBytes + FromBytes;
     type VerifierInput: ?Sized;
-    type VerifyingKey: Clone + ToBytes + FromBytes + From<Self::PreparedVerifyingKey> + From<Self::ProvingKey>;
+    type VerifyingKey: Clone
+        + ToBytes
+        + FromBytes
+        + Prepare<Self::PreparedVerifyingKey>
+        + From<Self::PreparedVerifyingKey>
+        + From<Self::ProvingKey>;
 
     fn setup<R: Rng>(
         circuit: &Self::Circuit,
         rng: &mut R,
-    ) -> Result<(Self::ProvingKey, Self::PreparedVerifyingKey), SNARKError>;
+    ) -> Result<(Self::ProvingKey, Self::VerifyingKey), SNARKError>;
 
     fn prove<R: Rng>(
         proving_key: &Self::ProvingKey,
         input_and_witness: &Self::AllocatedCircuit,
         rng: &mut R,
     ) -> Result<Self::Proof, SNARKError>;
 
-    fn verify(
-        verifying_key: &Self::PreparedVerifyingKey,
+    fn verify_with_processed_key(
+        processed_verifying_key: &Self::PreparedVerifyingKey,
```

Comment:
I think it may be more clear to call this `verify_prepared()` with `prepared_verifying_key`